ledger/src/blockstore.rs

```diff
@@ -1497,26 +1502,112 @@ impl Blockstore {
                 TransactionWithStatusMeta {
                     transaction: encoded_transaction,
                     meta: self
-                        .transaction_status_cf
-                        .get((slot, signature))
+                        .read_transaction_status((signature, slot))
                         .expect("Expect database get to succeed"),
                 }
             })
             .collect()
     }
 
+    fn initialize_transaction_status_index(&self) -> Result<()> {
+        self.transaction_status_index_cf
+            .put(0, &TransactionStatusIndexMeta::default())?;
+        self.transaction_status_index_cf
+            .put(1, &TransactionStatusIndexMeta::default())?;
+        // This dummy status improves compaction performance
+        self.transaction_status_cf.put(
+            (2, Signature::default(), 0),
+            &TransactionStatusMeta::default(),
+        )
+    }
+
+    fn toggle_transaction_status_index(
+        &self,
+        batch: &mut WriteBatch,
+        to_slot: Slot,
+    ) -> Result<Option<u64>> {
+        let index0 = self.transaction_status_index_cf.get(0)?;
+        if index0.is_none() {
+            return Ok(None);
+        }
+        let mut index0 = self.transaction_status_index_cf.get(0)?.unwrap();
+        let mut index1 = self.transaction_status_index_cf.get(1)?.unwrap();
+
+        if !index0.frozen && !index1.frozen {
+            index0.frozen = true;
+            batch.put::<cf::TransactionStatusIndex>(0, &index0)?;
+            Ok(None)
+        } else {
+            let result = if index0.frozen && to_slot >= index0.max_slot {
+                debug!("Pruning transaction index 0 at slot {}", index0.max_slot);
+                Some(0)
+            } else if index1.frozen && to_slot >= index1.max_slot {
+                debug!("Pruning transaction index 1 at slot {}", index1.max_slot);
+                Some(1)
+            } else {
+                None
+            };
+
+            if result.is_some() {
+                if index0.frozen {
+                    index0.max_slot = 0
```

Comment:
Seems like these writes to the index objects can race with writes to the same objects in `make_transaction_status_index`